### ISSUE TITLE
Retry policies for isolated (out-of-process) AF Cosmos DB triggers/bindings soon deprecated?

### ISSUE THREAD

**Issue by HMoen** [Azure/azure-webjobs-sdk-extensions#783]
@kshyju answer below works at the moment, but now we recently started to see the following trace in AppInsights for our AF Cosmos DB triggers :
"Soon retries will not be supported for function '[Function Name]'. For more information, please visit http://aka.ms/func-retry-policies."

The [Retry examples](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=fixed-delay%2Cisolated-process&pivots=programming-language-csharp#binding-error-codes) section also shows "Retry policies aren't yet supported when running in an isolated process." and the [Retries](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=fixed-delay%2Cisolated-process&pivots=programming-language-csharp#retries) section reflects no support for the Cosmos DB trigger/binding.

What's the path forward for AF Cosmos DB triggers running out-of-process?

Yes, retry policies are supported in isolated(out-of-process) function apps. You can enable it by adding the [retry section](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#retry) to your host config. Here is an example host.json
````
{
  "version": "2.0",
  "retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": 2,
    "delayInterval": "00:00:03"
  }
}
````

> The reason why I'm asking is the [documentation](https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=csharp#exponential-backoff-retry) mentioning
   Retries require NuGet package [Microsoft.Azure.WebJobs](https://www.nuget.org/packages/Microsoft.Azure.WebJobs) >= 3.0.23

That documentation which refers the usage of `ExponentialBackoffRetry` attribute is for in-proc function apps.

Please give it a try and let us know if you run into any problems.

_Originally posted by @kshyju in https://github.com/Azure/azure-functions-dotnet-worker/issues/832#issuecomment-1072545934_

**Comment by mattchenderson** [comment]
Because this is not exclusive to the isolated model, we should move this item to the extension repo. My understanding is that this would be a feature request for the Cosmos DB extension.

**Comment by MikeNicholls** [comment]
Will there be any guidance on retries for Cosmos DB triggered functions before the October 2022 deadline when retry policy support will be removed?

Cosmos DB is currently the outlier on the table of retry support for triggers/bindings, with "n/a" and "Not configurable" listed.
https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages#retries

**Comment by Ved2806** [comment]
Hi @ealsur Could you please help with this issue?

**Comment by iamarvil** [comment]
> Will there be any guidance on retries for Cosmos DB triggered functions before the October 2022 deadline when retry policy support will be removed?
> 
> Cosmos DB is currently the outlier on the table of retry support for triggers/bindings, with "n/a" and "Not configurable" listed. https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages#retries

I agree, we built our change feed processing with the retries in mind and we were just about to roll-out our solution. While moving to webjobs with our own change feed processor is a viable option, we would be losing other benefits that azure functions can provide. Any guidance for cosmos db change feed retries would be appreciated.

**Comment by ealsur** [comment]
At least at this time, there is no Retry Policy support on the Cosmos DB extension that is built in or uses any of those configurations

**Comment by HMoen** [comment]
@ealsur , any plans to implement before the October 2022 deadline mentioned above? If not, so we can plan our own retry mechanism (queue etc). Thanks.

**Comment by ealsur** [comment]
If I understand the notice correctly, the retry policy is being removed from all triggers that currently support it:

> The retry policy support in the runtime for triggers other than Timer and Event Hubs is being removed after this feature becomes generally available (GA). Preview retry policy support for all triggers other than Timer and Event Hubs will be removed in October 2022.

So adding Retry Policy support would not make sense as it does not align with the notice. 

We currently have no planned work to add custom retry configuration, but that can change if this is a feature that is widely requested (as an opt-in configuration). Technically, the Trigger could allow infinite retries or no retries (nothing intermediate) because we internally use the [Change Feed Processor](https://docs.microsoft.com/en-us/azure/cosmos-db/sql/change-feed-processor) and those would be the only available options.

**Comment by HMoen** [comment]
@ealsur , thanks for the feedback. Infinite or none won't work in our environment unfortunately. We piggybacked off [dead-letter-queues](https://docs.microsoft.com/en-us/azure/cosmos-db/sql/change-feed-processor?tabs=dotnet#error-handling) before and removed the functionality when we found out about the built-in retry mechanism. 

[Link](https://feedback.azure.com/d365community/idea/3bdd245e-841d-ed11-a81b-6045bd853198) added to upvote idea.

**Comment by mpaul31** [comment]
The removal of this has a very large impact on many of our production apps that use change feed (cosmos db trigger) with exponential infinite retries. We obviously don't want to lose and messages and rely heavily on this. @ealsur Is there anyway to get this added before the deadline or possibly a sample workaround? I would much rather not bring in more infra (queues) and build something specifically for this in all our apps.

**Comment by ealsur** [comment]
@mpaul31 I don't understand what is the request here. Right now the Change Feed Trigger has no retry policy, so it never retries, how are you creating apps that have exponential infinite retries?
The deadline that the documentation shows is a deadline after which the Retry Policies will be removed, for Cosmos DB they are not even there in the first place, so there is no point in adding them if they are going to be removed anyway?

This issue requires input from the Azure Functions team for clarification @Ved2806

**Comment by mpaul31** [comment]
@ealsur We use the following attribute on our change feed trigger method: 

`[ExponentialBackoffRetry(-1, "00:00:01", "00:00:10")]`

It would be awesome if we could configure similar behavior on the `CosmosDBTrigger`.

**Comment by ealsur** [comment]
There is no code in the Cosmos DB Trigger or extension that reacts to `[ExponentialBackoffRetry(-1, "00:00:01", "00:00:10")]`. 

The Trigger code in fact has no error retry logic, here is where your Function code is called: https://github.com/Azure/azure-webjobs-sdk-extensions/blob/dev/src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs#L225 if there was an execution error, it is simply logged and the next batch of changes is processed. This is mentioned in the Cosmos DB documentation (https://docs.microsoft.com/en-us/azure/cosmos-db/sql/troubleshoot-changefeed-functions#some-changes-are-missing-in-my-trigger).

There isn't (and never was) any logic that would pickup that `ExponentialBackoffRetry` decorator information in the Cosmos DB Trigger. If the Functions runtime is doing something with it I don't know, but the Trigger code moves forward always. My understanding is that the expectation that these RetryPolicy decorators had any impact on the CosmosDBTrigger are incorrect, so basically they are not doing anything, hence, when they get deprecated, there is nothing that will change.

**Comment by HMoen** [comment]
@ealsur , on this thread https://github.com/Azure/azure-functions-dotnet-worker/issues/832#issuecomment-1072545934, @kshyju mentioned the ExponentialBackoffRetry only works(ed) for in-proc function apps.

**Comment by ealsur** [comment]
I don't know how that works, from the Trigger perspective, there is nothing that is blocking the checkpointing and moving forward if the execution of the code fails.

Once the `ProcessChangesAsync` method completes in the Trigger code without errors, the state is saved and a next batch is obtained, and the execution of the user code can never make the method fail because we are using `TryExecuteAsync`:

```
FunctionResult result = await this._executor.TryExecuteAsync(new TriggeredFunctionData() { TriggerValue = docs }, cancellationToken);
```

For any logic to correctly affect the Trigger and make it retry the current batch (not checkpoint), the `ProcessChangesAsync` method should fail. That is how the Change Feed Processor would do the actual retry.

There is also no in-proc specific code in the `CosmosDBTriggerListener` either.

The linked comment is not saying that the configuration applies to CosmosDBTrigger either.

**Comment by HMoen** [comment]
The issue https://github.com/Azure/azure-functions-dotnet-worker/issues/832, was opened by someone looking specifically for retry logic for CosmosDBTrigger and @kshyju offered the linked comment as the solution. That issue should be updated to reflect your comments.

**Comment by ealsur** [comment]
I will let @kshyju comment then, as author of the CosmosDBTrigger, I am not aware of any behavior affected by these policies in the Trigger source code.

**Comment by kshyju** [comment]
> The issue [Azure/azure-functions-dotnet-worker#832](https://github.com/Azure/azure-functions-dotnet-worker/issues/832), was opened by someone looking specifically for retry logic for CosmosDBTrigger and @kshyju offered the linked comment as the solution. That issue should be updated to reflect your comments.

The retry option I mentioned is a function runtime feature which will retry the function execution when it fails for the first time

For example, if you have a function like this.

```
[Function("Function1")]
public void Run([CosmosDBTrigger(
    databaseName: "items-db",
    collectionName: "products-container",
    ConnectionStringSetting = "MyCosmosDbConnStr",
    LeaseCollectionName = "leases")] IReadOnlyList<MyDocument> input)
{
    if (input != null && input.Count > 0)
    {
        _logger.LogInformation($"Documents modified: {input.Count}");
        _logger.LogInformation($"First document Id: {input[0].Id}");

        // example to simulate failed function execution
        if (input[0].Name.Contains("foo"))
        {
            throw new ArgumentNullException("Bad name for  document");
        }
    }
}
```
and configure your `host.json` to enable retry behavior

```
{
  "version": "2.0",
  "retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": 2,
    "delayInterval": "00:00:03"
  }
}
```
Let' say you edit some document in your cosmosdb collection, the `Function1` function will be triggered and if the document name contains "foo", the function code simply throws, which creates a failed function execution. The function runtime(host) will retry the same trigger again after 3 seconds and this happens 2 times.

So in short, the retry behavior is something the functions runtime/host does based on what is configured on host.json. Hope this helps.

**Comment by HMoen** [comment]
@kshyju , thanks for your response. The retry behavior you depict above is/will be permanent in the runtime for out-of-proc functions?

I opened this issue since as stated above early June we started to see the following trace in AppInsights for our AF Cosmos DB triggers: "Soon retries will not be supported for function '[Function Name]'. For more information, please visit http://aka.ms/func-retry-policies."

**Comment by ealsur** [comment]
> The function runtime(host) will retry the same trigger again after 3 seconds and this happens 2 times.

Will the execution be with the same input (basically the same list of `IReadOnlyList<MyDocument> input`)? I'm intrigued to understand how this works, I have not added any logic in the Trigger that would stop the checkpointing, unless this decorator is basically stopping the instance?

**Comment by kshyju** [comment]
@ealsur  Yes, Same list. The logic is somewhere in the functions host I believe.

**Comment by kshyju** [comment]
> @kshyju , thanks for your response. The retry behavior you depict above is/will be permanent in the runtime for out-of-proc functions?
> 
> I opened this issue since as stated above early June we started to see the following trace in AppInsights for our AF Cosmos DB triggers: "Soon retries will not be supported for function '[Function Name]'. For more information, please visit http://aka.ms/func-retry-policies."

This is the current behavior as of today. Let me circle back with the team to understand what is the migration path & update you.

**Comment by mpaul31** [comment]
@kshyju Any updates on the migration path or postponing the removal?

**Comment by kshyju** [comment]
The retry behavior is currently a preview feature. As the deprecation notice mentions, this feature will only be available for Timer and EventHub triggers for GA.  Preview features are considered experimental and meant to collect feedback and not all of them make it to GA.

This will be applicable for both in-proc and isolated.

**Comment by mpaul31** [comment]
I get that but just like stage slots, which was in preview for a very (1+ years) long time, some adopt things sooner out of necessity. What about guidance then? Will the host just ignore the attribute or stop the app from starting? I'm sure there were many excited to see this as it relates to the cosmos trigger and are now going to be facing the same fate.

**Comment by shibayan** [comment]
The fundamental issue is not that `CosmosDBTrigger` cannot be configured with a retry policy, but that `CosmosDBTrigger` advances the lease when a Function fails to execute.

I believe this could be solved if `CosmosDBTrigger` instead of retry policy could use exponential backoff to extend polling time for Change Feed without advancing the lease on Functions failure.

**Comment by mpaul31** [comment]
Just to be clear, the retry policy/attribute does indeed work today aka with lease checkpointing working correctly. It appears it using an in memory retry policy similar to Polly.

**Comment by cheekp** [comment]
> I get that but just like stage slots, which was in preview for a very (1+ years) long time, some adopt things sooner out of necessity. What about guidance then? Will the host just ignore the attribute or stop the app from starting? I'm sure there were many excited to see this as it relates to the cosmos trigger and are now going to be facing the same fate.

Agree, I imagine this feature will have had a lot of adoption out of necessity over the 1-2 years of availability. With the loss of ExponentialBackoffRetry transient failures are going to become a lot more expensive again. Knock on effect to compute, source/target ops, logging etc.  Customers need guidance on what will happen in October and how to mitigate without the out-of-the-box in-memory retry option. Removing preview support for retry patterns and the alignment towards azure cloud best practices seems a strange decision.

**Comment by asos-martinsmith** [comment]
This is certainly required for Cosmos. 

Before I started using this attribute I encountered issues where "in flight" batches of changes were lost when performing a release. 

I was advised (funnily enough by @ealsur!) to use the retry policy to mitigate this here  

https://github.com/Azure/azure-webjobs-sdk-extensions/issues/716#issuecomment-824920182 

In general too the change feed processing may well be writing to things with 99.999% availability and there needs to be a way to handle the 0.001% without just skipping that batch and moving on.

**Comment by ealsur** [comment]
> I was advised (funnily enough by @ealsur!) to use the retry policy to mitigate this here

My response does not advise you to you the retry policies, the response is regarding what happens if your function crashes before checkpointing. I mentioned the documentation about the retry policies that were in preview because my understanding back then was that Functions was building a generic retry layer for all extensions, I did not explicitly say to use it. I don't work on the Functions team and have no insight on the roadmap nor plans for their features. I wasn't even aware these Retry policies affected the Cosmos DB Trigger (as you can see on this thread).

> In general too the change feed processing may well be writing to things with 99.999% availability and there needs to be a way to handle the 0.001% without just skipping that batch and moving on.

There is, a `try/catch` approach on the Function can let you handle any individual item failure. This pattern is commonly used (because retrying the entire batch is not something you would want to do in some cases), where the individual items that failed processing are sent to a queue or another storage for later processing or analyzing (why did they fail? is the failure transient?)

**Comment by asos-martinsmith** [comment]
There are two separate issues here. 

The first one is that certainly I have observed that deploying changes to the function app means that batches end up being aborted (maybe due to thread abort) and the lease is still updated so the net effect is that the batch is just missed.  After using this preview feature I did not observe any instances of this. When it is removed I no longer will have confidence in any method of deployment that avoids this, (it was never got to the bottom of this issue in the linked thread)

The second is that I don't want to provision a whole load of largely redundant resources for the 0.001% case (and write code to then merge from the fallback storage to the main storage). I just want the checkpoint not to be written and for it to be retried later (as happens with the exponential retry). Is there anything that can be put in the catch block that will achieve the suppression of checkpointing?

(Typical use case for me is to use the Change feed to upsert to a different Cosmos - e.g. to maintain a materialized view of a collection so pretty much all errors will be transient such as throttling and succeed on retry)

**Comment by ealsur** [comment]
> Is there anything that can be put in the catch block that will achieve the suppression of checkpointing?

Not at this point. As I mentioned in https://github.com/Azure/azure-webjobs-sdk-extensions/issues/783#issuecomment-1216727630, aligning with the Change Feed Processor, the only potential alternatives we could offer for the Cosmos DB Trigger is either an infinite retry (retry on any error) or no retry. There is no exponential backoff retry support on the Change Feed Processor. When checkpointing is aborted due to a processing error, the lease is released, any instance (same or other) could claim that and retry, so storing the number of previous attempts on an instance would not apply.

Unless the Retry Policies are exposed to Extension authors in a way that we could invoke/call/signal them in some way but the model from what I understand in this thread is that Retry Policies exist and apply outside of the Extension scope.

**Comment by shibayan** [comment]
Unless an alternative to the `CosmosDBTrigger` retry policy is provided by the Azure Functions team, we will unfortunately have to fix the Azure Functions Runtime version. Fixing `FUNCTIONS_EXTENSION_VERSION` to the version before the retry policy behavior was changed should give us additional time.

Would our request be better considered if we created a support request? I recall the removal of Azure Functions Proxies in the v4 upgrade. (BTW, Functions Proxies are about to be put back after being removed)

**Comment by iamarvil** [comment]
As a temporary fix, we'd be wrapping our triggers with Polly in the meantime just to cover transient errors. Consumption plan timeouts would be the next thing to worry about, but considering that processing usually takes less than a second anyway, i don't think it's too much of a concern. We're just unsure how much cost it will add to our azure functions bill.

I would consider adding a try catch block to send erring processes, to queues but we really rely on the order of documents coming through.

A retry forever with a fixed wait duration from the cosmosdb trigger would be great though, and is exactly what we need.

**Comment by sayyaari** [comment]
@ealsur is retry policy support for Cosmos DB Trigger that you've added recently an intermittent change until October 2022 according to this [document](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=fixed-delay%2Cin-process&pivots=programming-language-csharp#handling-errors) or the document just needs to  be updated?

**Comment by ealsur** [comment]
@sayyaari The change I added is for the 4.x extension that will release sometime next month, the document would need to be updated I guess

**Comment by sayyaari** [comment]
Thanks @ealsur, Still we have a major problem which is when the Azure Function is terminated by any reason it would advance the lease. Per my investigation in the code it is because the **ProcessChangesAsync** method of [CosmosDBTriggerListener](https://github.com/Azure/azure-webjobs-sdk-extensions/blob/dev/src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs) which handles the **ChangeFeedProcessor** doesn't throw any exception so that the lease doesn't advance in case of such situation. It just logs when get failed results and progress gracefully. Is there any plan to change this behaviour and make it for example more configurable for the next release?

**Comment by ealsur** [comment]
@sayyaari I believe you should create a different issue that we can discuss separately. Azure Function being terminated and the Listener not aborting on errors are two different things. Please see: https://github.com/Azure/azure-webjobs-sdk-extensions/issues/716#issuecomment-824910846

**Comment by mpaul31** [comment]
> Unless an alternative to the `CosmosDBTrigger` retry policy is provided by the Azure Functions team, we will unfortunately have to fix the Azure Functions Runtime version. Fixing `FUNCTIONS_EXTENSION_VERSION` to the version before the retry policy behavior was changed should give us additional time.
> 
> Would our request be better considered if we created a support request? I recall the removal of Azure Functions Proxies in the v4 upgrade. (BTW, Functions Proxies are about to be put back after being removed)

@shibayan Is pinning the version the route you are going with? I am debating this approach vs. adding a Polly retry policy.

**Comment by shibayan** [comment]
@mpaul31 I am currently looking at ways to pin Azure Functions Runtime for safety and then support the new Cosmos DB Extension v4. The reason is to avoid losing retry support for `CosmosDBTrigger` in Azure Function Runtime due to platform updates before the Cosmos DB Extension v4 migration is complete.

The Azure Functions team has disclosed that the previous retry policy will be removed in October, so the risk is that we cannot read when exactly in October the removal will take place.

**Comment by mpaul31** [comment]
> @mpaul31 I am currently looking at ways to pin Azure Functions Runtime for safety and then support the new Cosmos DB Extension v4. The reason is to avoid losing retry support for `CosmosDBTrigger` in Azure Function Runtime due to platform updates before the Cosmos DB Extension v4 migration is complete.
> 
> The Azure Functions team has disclosed that the previous retry policy will be removed in October, so the risk is that we cannot read when exactly in October the removal will take place.

I am in the exact same boat. Updating the app setting `FUNCTIONS_EXTENSION_VERSION` value to version `3.15.0.0`  seems less invasive at this point.

**Comment by Ved2806** [comment]
Hi @HMoen Is this resolved for you?

**Comment by HMoen** [comment]
> The Azure Functions team has disclosed that the previous retry policy will be removed in October, so the risk is that we cannot read when exactly in October the removal will take place.

@Ved2806 , has this been confirmed?

**Comment by Ved2806** [comment]
Hi @shibayan Could you please confirm?

**Comment by shibayan** [comment]
@Ved2806 I believe it will be difficult to confirm until the new Cosmos DB Extension v4 goes GA.

The Cosmos DB Extension v4 should reach GA before the retries in Cosmos DB Extension v3 are removed from Azure Functions Runtime.
https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.CosmosDB/

**Comment by ealsur** [comment]
For what it's worth, the `rc` package contains the code for the Retry Policy integration: https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.CosmosDB/4.0.0-rc

**Comment by shibayan** [comment]
Retry policy works in v4.0.0-rc, but the scale controller no longer monitors Change Feed when used with Consumption Plan.

As a result, when a change occurs in the Change Feed, Azure Functions are not started and processing is not performed.

**Comment by ealsur** [comment]
@shibayan That is the reason the package is not yet GA, Scale Controller support is not completed deployment yet (reference https://azure.microsoft.com/en-us/updates/azure-cosmos-db-azure-functions-extension-version-40-in-public-preview/)

**Comment by iamarvil** [comment]
It's finally on GA! Thank you very much @ealsur!

We can finally update now!

**Comment by HMoen** [comment]
@shibayan @ealsur , does this enhancement apply to out-of-proc functions as well or [docs](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-error-pages?tabs=exponential-backoff%2Cisolated-process&pivots=programming-language-csharp#retry-examples) are accurate?

![image](https://user-images.githubusercontent.com/2124254/220148587-40427f35-c7ce-424d-84bb-79793f717a29.png)

**Comment by ealsur** [comment]
@HMoen I cannot comment, I am not aware of how out-of-proc works, but based on the documentation you linked, there is no support for any extension.

**Comment by HMoen** [comment]
Thanks for the feedback. I opened this issue specifically for out-of-proc functions, so will await the response of @shibayan or @Ved2806  before closing.

**Comment by ghost** [comment]
Is there any update on retry policies for CosmosDB triggers on out-of-process functions?
.NET 7+ supports only out-of-process functions and without retry policies CosmosDB trigger is unusable because it will keep losing changes.

Perhaps this question should be taken to azure-functions-dotnet-worker repository?

**Comment by joeizy** [comment]
# Overview
Hello, after some extensive testing with the latest version of nuget packages for out-of-process dotnet cosmos change feed function here's what I discovered the behavior to be. PLEASE correct me where I missed something 😄 

**Goal**: My goal is to be able to use Azure Functions with Cosmos Change Feed as a way to reliably process changes coming from Cosmos DB. (Think replication to another collection, notifying consumers, etc.)

**Problem**: I found that the behavior of code written in the way one might expect to write it (i.e. throw exception in change feed processor function) still advances the change feed in edge cases making the provided tooling for Azure Function Change Feed Trigger into a "best effort" process instead of reliable processing (see scenarios below). This behavior means that changes in the Change Feed can be lost unexpectedly (to the developer writing code) in edge cases.

**Environment**
```
func --version
4.0.5198

dotnet --version
7.0.304

*.csproj
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <TargetFramework>net7.0</TargetFramework>
        <AzureFunctionsVersion>V4</AzureFunctionsVersion>
        <OutputType>Exe</OutputType>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="Azure.Storage.Queues" Version="12.14.0" />
        <PackageReference Include="Microsoft.Azure.Functions.Worker" Version="1.14.1" />
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.CosmosDB" Version="4.3.0" />
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Http" Version="3.0.13" />
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Extensions.Storage.Queues" Version="5.1.2" />
        <PackageReference Include="Microsoft.Azure.Functions.Worker.Sdk" Version="1.10.0" />
    </ItemGroup>
...
</Project>

OS
Edition	Windows 11 Pro
Version	22H2
Installed on	‎10/‎15/‎2022
OS build	22621.1848
Experience	Windows Feature Experience Pack 1000.22642.1000.0
```

The failure conditions / edge cases I tried are meant to simulate:
- The function runtime process crashing
- The isolated worker process crashing
- A dependency of the user function code being down for an extended period of time (i.e. a service outage)
- The user function code having a bug requiring a code deployment (wherein the function runtime is gracefully shutdown)

NOTE: I did these tests with the `[FixedDelayRetry]` attribute on my function. It's my understanding (and assumption) that this behavior is similar if not worse with out the retry attribute.

# Findings

NOTE:
When I say "runtime process" I'm referring to `func.exe` from the Azure Function Core Tools locally.
When I say "worker process" or "isolated process" I'm referring to `dotnet.exe` running my user function code locally.

The change feed _will still advance_ in all of these cases:
- Unhandled Exceptions in User Function Code will still advance the change feed
- Add `ExponentialBackoffRetry[]` or `FixedDelayRetry[]` attributes will retry in-memory but if the runtime process is gracefully stopped (i.e. code deployment, environment restart, etc.) it will still advance the change feed
- Repeated throwing exception in Function will still advance the change feed
- Repeated crashing the worker process (i.e. with `Environment.FailFast()` in the User Function Code) will still advance the change feed
- Exiting the worker process with a non-zero status code (i.e. `Environment.Exit(-1)`) seems that it will hang indefinitely but when you go to stop the runtime process, it will advance the change feed.


The change feed _will not advance_ in any of these cases:
- The runtime process is killed or terminates unexpectedly (i.e. host VM crashed, Process is terminated via task manager, etc.). 
- The runtime process starts successfully, but the worker process does not start successfully. (i.e. The worker process crashes before `host.Run()`)
- The runtime and worker processes start and begin to process a message but then the worker (isolated) process crashes and the auto-restart of the worker process (which is controlled by the runtime process) exceeds the number of retries (i.e. the worker process crashed and never successfully started) with this **Error Message:** `Exceeded language worker restart retry count for runtime:dotnet-isolated. Shutting down and proactively recycling the Functions Host to recover`. This crash of the worker process during startup must take place **before** the `host.Run()` in `Program.cs`. NOTE: If the worker process successfully restarts (ie. `host.Run()` succeeds) it doesn't matter that you continue to crash the process in the function code, the feed will still advance; you must crash the worker process before `host.Run()`.

**NOTE**: I did find it unusual that the runtime Cosmos trigger so was lenient with when it advances the change feed lease/tracker and expected it to be a lot more conservative. For example, even when I see this in the logs which clearly indicates a scenario where the worker is not happy during processing the change feed tracking is still advancing.

```
Language Worker Process exited. Pid=43372.
Exceeded language worker restart retry count for runtime:dotnet-isolated. Shutting down and proactively recycling the Functions Host to recover
Lease 0 encountered an unhandled user exception during processing.
```

# How to get reliable processing (workaround)
The only thing I was able to get to work was to use `Environment.FailFast()` in the User Function Code when encountering an issue that is not recoverable to kill the isolated worker processed. I **must** crash the worker process, simply throwing exceptions in user code or even `Environment.Exit(1)` doesn't work here. 

Then, when the runtime process restarts the worker process, I need to detect that failure condition in the bootstrap / startup code (inside the `Program.cs`) of the worker process and crash the process before `host.Run()` is called. Crashing here can be throwing unhandled exception or `Environment.FailFast()` or even `Environment.Exit([non-zero value])` with non-zero exit code it doesn't matter as long as `host.Run()` isn't called and the process exits in a failure condition.

In this scenario, the runtime process will restart the worker process a few times then give up with error message `Exceeded language worker restart retry count for runtime:dotnet-isolated. Shutting down and proactively recycling the Functions Host to recover` and the runtime process will exit. The Change Feed is not advanced and the next time the runtime process is started it will attempt to process the change in the feed again.

**NOTE**: "Just putting the message on a queue from code in your change feed function" (i.e. ServiceBus or Storage) won't work here because that operation can also fail and if the current in-progress change is not properly retried it will be lost. Even in the "put it on a queue" scenario, we need to reliably get the value from change feed to the queue without dropping changes. Due to the above edge cases that isn't possible without this workaround from that I can see.

**Comment by Narvalex** [comment]
hey @joeizy , how would you exactly detect in `Program.cs` that you crashed in the previous run?

**Comment by joeizy** [comment]
Hello @Narvalex, I was not detecting a previous failure (IMO this is not productive), I was predicting whether the next attempt to process the message would knowing fail (i.e. does it have the prerequisites/dependencies to succeed). The assumption is that the User Function Code is idempotent AND that it will not fail unless there is a dependency which is failing or inaccessible. So, if I can check whether the dependencies are working I'm set.

Consider a case where the User Function Code receives the change feed item and puts a message on a durable queue.

If the durable queue is down or the function is misconfigured with an incorrect connection string or there is misconfigured networking between the function and the queue, the attempt to enqueue the message will fail. This can be checked before `host.Run()` is called during startup (for example by putting a dummy type message).

This is NOT perfect as I'm sure you can imagine other scenarios where things might fail which you cannot proactively check for but it is quite common for a dependency failure to be the cause (and one of the main uncontrollable reasons) and it's common to want to be resilient to this case (i.e. not lose / drop messages due to intermittent failures).

NOTE: It's my understanding that further resilience requires a change to the Cosmos Trigger Code (which controls advancement to the next change in the feed) to adopt an at-least-once-successful message delivery pattern with dead-letter queue (which I believe most people are expecting it already does) OR to delegate advancement of the cursor / change feed to the User Function Code (advanced use cases).
